_source/_posts/2019-05-07-nodejs-typescript-api.md

```diff
@@ -732,30 +739,26 @@ ReactDOM.render(
 );
 ```
 
-You can create a new React hook to help with authentication as well. This will require you to pass in an `auth` variable, which will then be used to determine whether or not a user is authenticated, find out info about the user, and get the access token. These are then passed back into your React component for user later. Create a new file `src/client/auth.ts`:
+You can create a new React hook to help with authentication as well. You can use the Okta React SDK's `useOktaAuth()` hook to determine whether or not a user is authenticated, find out info about the user, and get the access token. These are then passed back into your React component for user later. Create a new file `src/client/auth.ts`:
 
 ```typescript
 import { useEffect, useState } from "react";
+import { useOktaAuth } from "@okta/okta-react";
 
-export const useAuth = (auth) => {
+export const useAuth = () => {
+  const { authService, authState } = useOktaAuth();
   const [authenticated, setAuthenticated] = useState(null);
   const [user, setUser] = useState(null);
   const [token, setToken] = useState(null);
 
   useEffect(() => {
-    auth.isAuthenticated().then((isAuthenticated) => {
-      if (isAuthenticated !== authenticated) {
-        setAuthenticated(isAuthenticated);
-      }
-    });
+    setAuthenticated(authState.isAuthenticated)
   });
 
   useEffect(() => {
     if (authenticated) {
-      auth.getUser().then(setUser);
-      auth.getAccessToken().then((accessToken) => {
-        setToken(accessToken ? `Bearer ${accessToken}` : null);
-      });
+      authService.getUser().then(setUser);
```

Comment:
You're calling getUser(), which is asynchronous, but this entire hook is synchronously returning `user`, so `user` hasn't been set yet.

This same sort of issue is why useOktaAuth() returns the .isAuthenticated flag and a state object - so they can wait until it a token retrieval is complete before trying to use the token.  (We actually are looking at adding userinfo to that object, but didn't want to commit everyone to the extra call unless they needed it).

Reply:
Is there any sort of workaround? I tried wrapping the call in async/await, but that doesn't seem to work.

Reply:
I got it working! Here's the code:

```js
import { useEffect, useState } from "react";
import { useOktaAuth } from "@okta/okta-react";

export const useAuth = () => {
  const { authService, authState } = useOktaAuth();
  const [user, setUser] = useState(null);
  const [token, setToken] = useState(null);

  useEffect(() => {
    if (authState.isAuthenticated) {
      authService.getUser().then(setUser);
      setToken(`Bearer ${authState.accessToken}`);
    } else {
      setUser(null);
      setToken(null);
    }
  });

  return [user, token];
};
```

Reply:
yeah, useEffect itself is synchronous. 

You can return a function that returns an enclosed value instead of a value, or....

Actually, there's another issue here - this runs on every render, so they will be calling /userinfo every render. 

I would remove useEffect() here and wrap the call to `.getUser()` in a `if(!user)` - then the calls to setState() should do what you want, and trigger on a rerender.  Let me know if that works.

Reply:
Addendum: useState() is ALSO synchronous, but the call to setState() should trigger a new call, which will return the actual value of user, since setting the user is what triggered the re-render.

Reply:
This seems to work:

```js
  useEffect(() => {
    if (authState.isAuthenticated) {
      if (!user) {
        authService.getUser().then(setUser);
      }
      setToken(`Bearer ${authState.accessToken}`);
    } else {
      setUser(null);
      setToken(null);
    }
  });
```